docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,793 @@
+poor “Separation of Concerns” while designing such components.
+
+### Deploy Runtime Extensions
+Cluster API requires that each Runtime Extension must be deployed using an endpoint accessible from the Cluster API
+controllers; additionally, Runtime Extensions must always be executed out of process (not in the same process as
+the Cluster API runtime).
+
+This proposal assumes that the default solution that implementers are going to adopt is to deploy Runtime Extension
+in the Management Cluster by:
+
+- Packing the Runtime Extension in a container image;
+- Using a Kubernetes Deployment to run the above container inside the Management Cluster;
+- Using a Cluster IP service to make the Runtime Extension instances accessible via a stable DNS name;
+- Using a cert-manager generated Certificate to protect the endpoint.
+
+There are a set of important guidelines that must be considered while choosing the deployment method:
+
+#### Availability
+It is recommended that Runtime Extensions should leverage some form of load-balancing, to provide high availability
+and performance benefits; you can run multiple Runtime Extension backends behind a service to leverage the
+load-balancing that services support.
+
+#### Identity and access management
+The security model for each Runtime Extensions should be carefully defined, like for any other application deployed
+in the Cluster: the deployment must use a dedicated service account with limited RBAC permission.
+
+On top of that, the container image for the Runtime Extension should be carefully designed in order to avoid
+privilege escalation (e.g using [distroless](https://github.com/GoogleContainerTools/distroless) base images) and
+the Pod spec in the Deployment manifest should enforce security best practices (e.g. do not use privileged pods etc.).
+
+####  Alternative deployments methods
+Alternative deployment methods can be used given that the requirement about the HTTP endpoint accessibility
+is satisfied, like e.g.
+
+- deploying the HTTP Server as a part of another component, e.g. a controller;
+- deploying the HTTP Server outside of the Management Cluster.
+
+- Also in this case above recommendations about availability and identity and access management apply.
+
+### Registering Runtime Extensions
+
+_Important! Cluster administrators should carefully vet any Runtime Extension registration, thus preventing
+malicious components from being added to the system._
+
+_Creating RuntimeExtensionConfiguration will be allowed only if the Runtime Extension feature flag is set to true._
+
+By registering a Runtime Extension the Cluster API Runtime became aware of a Runtime Extension implementing a
+Runtime Hook, and as a consequence the runtime starts calling the extension at well-defined moments of the
+Workload Cluster’s lifecycle.
+
+This process has many similarities with registering dynamic webhooks in Kubernetes, but in this case it is required
+to use the following CR:
+
```

Comment:
Is this an example with pseudo API for reference or this proposal including a new RuntimeExtensionConfiguration API kind as it looks below?

Reply:
This proposal includes a new RuntimeExtensionConfiguration kind as it looks below +/- improvements if we get agreement